service/share/backoff.go

```diff
@@ -0,0 +1,70 @@
+package share
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p/p2p/discovery/backoff"
+)
+
+// backoffConnector wraps a libp2p.Host to establish a connection with peers
+// with adding a delay for the next connection attempt.
+type backoffConnector struct {
+	h       host.Host
+	backoff backoff.BackoffFactory
+
+	cacheLK      sync.Mutex
+	backoffCache map[peer.ID]connectionCacheData
+}
+
+// connectionCacheData stores time when next connection attempt with the remote peer.
+type connectionCacheData struct {
+	nexttry time.Time
+	backoff backoff.BackoffStrategy
+}
+
+func newBackoffConnector(h host.Host, factory backoff.BackoffFactory) *backoffConnector {
+	return &backoffConnector{
+		h:            h,
+		backoff:      factory,
+		backoffCache: make(map[peer.ID]connectionCacheData),
+	}
+}
+
+// Connect puts peer to the backoffCache and tries to establish a connection with it.
+func (b *backoffConnector) Connect(ctx context.Context, p peer.AddrInfo) error {
+	b.cacheLK.Lock()
+	cacheData, ok := b.backoffCache[p.ID]
+	strategy := b.backoff()
```

Comment:
nit: don't really understand the grouping here and why ok check is not close to its definition/read from map

Reply:
Yes, I made wrong assumption. Will fix it.

Reply:
I was added this way to be able to assign it in one place after `if ok` condition for both cases.